File: ParameterPassingMechanisms/CallByResult.cpp
/* Ryan Hirscher
 * Dr. Oudshoorn
 * CSC4510
 * Assignment 2 Call By Result C++
 *
 * This program attempts to demonstrate C++ pass by result  mechanism and scope with a swap function and printing out the values and addresses of the actual and formal variables.
 *
 * C++ does not support out exclusive semantics for parameters, only pass by reference and pass by value. It does not make sense for implementation as it is not a supported feature. However, the feature can be mimiced by passing with & or * with a dummy variable. Since we are referencing the memory, z can be returned while passing in just an empty dummy pointer or address reference.
 * 
*/

#include<iostream>
using namespace std;
void swap (int, int, int&);
int main()
{
   int x=10;
   int y=999;
   int z;
   //What is the address of the actual values? - z is the same as formal, x and y are not the same
   cout << "Call by result to swap function" << endl;
   cout << "Pre-swap actual X: " << x << endl;
   cout << "X memory address: " << &x << endl;
   cout << "Pre-swap actual Y: " << y << endl;
   cout << "Y memory address: " << &y << endl << endl;
   
   //Passing the actual
   swap(x, y, z);

   //Did the memory location of the actual values change? - x and y changed, z did not
   //Did the values change from the function? - The swap did not retain, z actual and formal retained
   cout << "Post-swap actual X: " << x << endl;
   cout << "X memory address: " << &x << endl;
   cout << "Post-swap actual Y: " << y << endl;
   cout << "Y memory address: " << &y << endl;
   cout << "Post-swap actual Z: " << z << endl;
   cout << "Z memory address: " << &z << endl << endl;
   
   return 0;
}

void swap (int x, int y, int &z)
{
   int temp=0;
   temp = x;
   x = y;
   y = temp;
   
   z = x + y;
   //Do x and y retain the same memory locations? - z had the same location, but x and y changed
   //Did the values swap? - They did in local scope
   cout << "In-swap formal X: " << x << endl;
   cout << "X memory address: " << &x << endl;
   cout << "In-swap formal Y: " << y << endl;
   cout << "Y memory address: " << &y << endl;
   cout << "In-swap formal Z: " << z << endl;
   cout << "Z memory address: " << &z << endl << endl;

   return;
}

File: ParameterPassingMechanisms/CallByResult.java
/* Ryan Hirscher
 * Dr. Oudshoorn
 * CSC4510
 * Assignment 2 Call By Result Java
 *
 * Java is sctrictly built to support pass by value. Addresses are not able to be retrieved for variables.
 *
 * Java does not support Call by Result. However, it can be implmeneted to accomplish similar behavior through defining your own class objects. These objects are referenced through pointers under the hood, so changes made to the formal parameters will be reflected back to the actuals even if you were to change the name of the formal parameters. So passing a dummy variable and assigning it new value at the return would reflect how Call By Result should function.
 *
 * Actual value changes are reflected in main.
*/

//Callee and defining object class CallByRef
class CallByR {
   int X, Y;

   //Method for assignment
   CallByR(int x, int y)
   {
      X=x;
      Y=y;
   }

   //Method to add 
   static void addFive(CallByR obj, CallByR res)
   {
         obj.X+=5;
         obj.Y+=5;
         System.out.println("Formals: X=" + obj.X + " Y=" + obj.Y);
         System.out.println("Formals Result: X=" + res.X + " Y=" + res.Y);

         res.X=obj.X;
         res.Y=obj.Y;
         return;
   }
}

//Caller
class CallByResult {
   public static void main(String[] args)
   {
      //Declaring object type CallByR named objectActual and objectResult
      //New instance and storing 5 as X and 12 as Y and dummy object with 0s
      CallByR objectActual = new CallByR(5, 12);
      CallByR objectResult = new CallByR(0,0);

      System.out.println("Call By Result (class object) to addFive function.");
      System.out.println("Actuals Pre-add: X=" + objectActual.X + " Y=" + objectActual.Y);
      System.out.println("Actuals Result: X=" + objectResult.X + " Y=" + objectResult.Y);

      //Calling
      objectActual.addFive(objectActual, objectResult);

      System.out.println("Actuals Post-add: X=" + objectActual.X + " Y=" + objectActual.Y);
      System.out.println("Post-add Result: X=" + objectResult.X + " Y=" + objectResult.Y);
   }
}
